### ISSUE TITLE
Paperwork 0.2 UX issues

### ISSUE THREAD

**Issue by mjourdan** [jflesch/paperwork#356]
Hi,

While I had to scan about a dozen of documents, I gave a try to Paperwork and would like to share some thoughts on it's user experience. You'll find in a separate post some ideas to fix the issues listed below.

Ux review
=========

First things first, everytime I launch Paperwork, I'm delighted by it's logo and think "straight and fast, this will light me of a burden". However, Paperwork is not as close to the "scan and forget" philosophy as I hoped, because reaching the scan step requires too much effort and time (see how lazy I am).

In my opinion, controls make user think too much :

0. huge number of buttons make feel every actions are equally important. This should be "scan and forget", but it more looks like "scan or modify or print or move or zoom or open a dir and forget"
1. document controls (new, edit, delete) are separated from the document list (by a too wide searchbar)
2. useless quit button
3. the app menu contains actions specific to the current window
4. Settings icon and Preferences in the app menu do exactly the same thing but appear twice and under different names

I also think the actual content is drawn in a far too complex layout:

5. tabs to switch between labels and pages make think labels are as important as pages
6. at a time, the document list shows no more than 2 documents (on a 15" screen)
7. 3 columns layout for two kinds of items (documents and pages), so a single page may appear three times at a time (two times as thumbnails in document list and pages list, plus the page itself)
8. some empty space at the bottom of the screen look very weird
9. huge amount of vertical space lost (up to 7 lines of controls + windows decoration on the labels tabs)

Once she have figured out how the ui works, user will need to find out how to get this stupid scanner working: 

10. scanner detection requires an extra step on every launch (Hint: open preferences, wait a couple of seconds until a scanner is found, then close the preferences)

Finally, it's pretty easy to get rid of a maximum of paper in a minimum of time. Being able to move pages across documents and marking documents with labels are great features, for example to avoid bothering with scaning the right pages in the right order. Nonetheless, there is plenty of room for improvement.

Regards

**Comment by jflesch** [comment]
I have a problem: In your mockups, you use 2 [header bars](https://developer.gnome.org/gtk3/stable/GtkHeaderBar.html), and it doesn't work.

![Image of what happens when we try to have 2 header bars](http://jflesch.kwain.net/~jflesch/paperwork/paperwork_crap.png).

As you can see, there are visual issues (I can take care of the height of the header bars, but not their corners). Also, because we want two of them, I cannot use [GtkWindow->set_titlebar()](https://developer.gnome.org/gtk3/stable/GtkWindow.html#gtk-window-set-titlebar), which means I have to disable decorations to make the usual title bar disappear ... which means the window cannot be resized anymore.

I will try and see if I can pack GtkLabels in the header bar. In which case, it could help solve most of this problem. However, I will have to find a way to make a clean separation between the 2 "fake" header bars ... maybe a custom Gtk widget.

**Comment by jflesch** [comment]
Note to myself:

The XML for the header bar (Glade doesn't support header bars yet, so I will have to make it by hand):
```
<object class="GtkHeaderBar" id="headerbar">
  <property name="visible">True</property>
  <property name="title">Paperwork</property>
  <property name="subtitle">Grep for dead tres</property>
  <property name="show_close_button">False</property>
</object>   
```

**Comment by MathieuJourdan** [comment]
Some other pieces of software already use left pane since gtk 3.10. You may take a look at [gnome-contacts](https://github.com/GNOME/gnome-contacts), [gedit](https://github.com/GNOME/gedit) or [polari](https://wiki.gnome.org/Apps/Polari).

You may want to refer to the latest mockups on my repo, in particular one to [browse papers by label](https://github.com/MathieuJourdan/gnome-design/blob/master/paperwork/paperwork-label-filter.png).

**Comment by jflesch** [comment]
You're right.
From what I can see in Polari, the trick I missed is that the 2 GtkHeaderBar must be in the same parent GtkBox. I'm still a little bit fuzzy regarding how they keep the length of the GtkHeaderBar consistent with the rest of the window, but that's easy to manage anyway.

**Comment by jflesch** [comment]
Funny thing, Polari use GtkWindow->set_titlebar() but not Gnome-contacts. Meh, whatever works.

**Comment by jflesch** [comment]
Note to myself:
For the document list : https://developer.gnome.org/gtk3/stable/GtkListBox.html
(to check: if the performances are ok ... otherwise, a custom widget may be required :/)

**Comment by MathieuJourdan** [comment]
To keep width consistent, Gedit seems to put its side_panel and side_headerbar in a [GtkSizeGroup](https://developer.gnome.org/gtk3/stable/GtkSizeGroup.html) in its [ui file](https://github.com/GNOME/gedit/blob/e4715b56866106958cbde5e3c0899a5a7781b9bd/gedit/resources/ui/gedit-window.ui) and bind them together in its [window.c](https://github.com/GNOME/gedit/blob/aa7c9af6865faade92a74b77813442d2420f277c/gedit/gedit-window.c). Contacts seems to do [pretty much](https://github.com/GNOME/gnome-contacts/blob/master/data/ui/contacts-window.ui) the [same thing](https://github.com/GNOME/gnome-contacts/blob/271d3a77cc82dedfd215bfb740409c0976124fcf/src/contacts-window.vala).

**Comment by jflesch** [comment]
Nice, I didn't know about GtkSizeGroups.

**Comment by jflesch** [comment]
Ok, got it : 09bd22672896caba9f390485380db32a90d43bf0 + adc9238880323a5d26c32f3ed9cbbe41188e2e1b (branch: wip-ui-redesign)

**Comment by jflesch** [comment]
Header bar content: 00aa327850740c1afd369a304400805c68809536 (needs some minor style improvement + there is a workaround I should remove)

**Comment by MathieuJourdan** [comment]
Nice to see things going on!

**Comment by jflesch** [comment]
Scraping off the status bar, the page list and the label list: ad3df2d1391b11a23de49bcbff53eab717c0056a

**Comment by jflesch** [comment]
ca71a83d387d8a6db55f881ffb70560088d37e1c : Fix the style of the page numbers

**Comment by jflesch** [comment]
Current state : ![img](http://jflesch.kwain.net/~jflesch/paperwork/paperwork_ui_redesign_progress.png)

**Comment by MathieuJourdan** [comment]
Some Feedback here, while it is still wip.

*New document button:* icon looks perfectly relevant but I worry it may confuse users. The page symbolic should be used either to represent documents, either to represent pages. If we chose to keep the image_button_new_doc, then we should find something else to switch between "single page" and "grid" in the content view.

*App menu:* having Edit/Export/Print/Delete entries duplicated under two top-level menu feels weird. The vast majority of software first offers to select Export/Print, and then offers to either proceed with the whole document or to pick up some pages. Also, if editing a page is rotating it or croppring it, I can't figure out what editing a document could be. Probably we should be more specific here, e.g. use "document properties" instead of "edit".

*Content view:* scrollbars should be hidden when there is nothing to scroll.

**Comment by jflesch** [comment]
1) New document button + icon : Fair enough. I can change it back to a simple '+'. IMHO, it doesn't matter that much anyway.

2)
 a) Hm, 'print' shouldn't ac be duplicated. My bad.
 b) No problem with renaming 'Edit' into 'Properties' for the documents
 c) I think 'delete' can safely be duplicated as of now. IMHO, it's easy to understand.
 d) Regarding the export options and dialog, if you have any ideas / mockups, please feel free to share them. But right now, I think the way it's done currently is probably the simpler.

(Note: as you pointed out yourself in your initial UX review, it's not the application menu. I guess the correct name should be 'window menu' :)

3) Regarding scroll bars, I will see what I can do.

**Comment by MathieuJourdan** [comment]
Regarding the second point:

a. I think we shoud avoid duplication here, we can make simpler things
b. Do you think the "properties" and "label" buttons (bottom-left toolbar on the mockup) are an unappropriate way to go? Would you be interested with mockups of a popover on the document title in the headerbar?
c. What about making the bin a larger area (let say a third of the left pane width), so that both documents and pages could be dragged and dropped to?
d. I'll see what I can do, the export is not that usable for now

**Comment by jflesch** [comment]
b. No, I actually forgot about these buttons in the bottom-left when I made the menus .... I'll remove 'properties' from the menu ... :)

c. The bin ? For deletion you mean ? No, deletion is not an operation that much common. I don't think a larger area.

Hm, here are some random ideas:
1) For pages, the buttons 'delete' and 'edit' can be put on the grid view (a pencil at the bottom left of each thumbnail, and a 'X' at the bottom right of each)
2) In the window, only rarely used options like : 'open directory', 'redo ocr', 'print', and 'export'. 'export' would then be a submenu to specify 'document' or 'page'
I think with that, we would have all the options placed in a smart way. What do you think ?

**Comment by jflesch** [comment]
Actually, I think we can apply the same logic everywhere : there is no need for a toolbar at the bottom.
For each document, I can add buttons "properties", "labels", and "delete".

As for the page grid, the button will only be display when the mouse goes over the document (or the page)

By the way, I also think it would make sense to merge the label and properties dialogs (--> one single button).

**Comment by MathieuJourdan** [comment]
> Actually, I think we can apply the same logic everywhere : there is no need for a toolbar at the bottom.
For each document, I can add buttons "properties", "labels", and "delete".

I thought of the toolbar to avoid cluttering the document pane, but I agree that wouldn't work as well as I hoped.

> As for the page grid, the button will only be display when the mouse goes over the document (or the page)

How would it work with touch screens ? Maybe a "delete document" in the window menu would fit better?

> By the way, I also think it would make sense to merge the label and properties dialogs (--> one single button).

It makes perfect sense; please find a mockup below.

1. Left pane:
    1. stuff is called "papers" instead of "documents", to reflect Paperwork's scope (documents like slideshows and spreadsheets are somewhat out of scope)
    2. papers can be sorted either by date or by title (should probably be hidden and auto-reveal when reaching the top of the list)
    3. papers can be deleted using a specific icon (doesn't look nice, I think)
2. Document properties:
    1. the first text field is expandable, the first line will be used as a title for the paper, others are extra words. 
    2. the second field allows to change the date
    3. the third field offers self-completion to easily pick up labels

![paperwork-properties](https://cloud.githubusercontent.com/assets/5346620/5789399/19678778-9e69-11e4-82ef-8726914436cf.png)

**Comment by jflesch** [comment]
Hm, for me, the window menu should contains options rarely used. From my experience, those options are used quite often:
* doc properties : because label guessing doesn't always work as well as I would it to :/
* page edit : because page orientation screw up sometimes
* page delete / doc delete : because sometimes you're tired, you put the sheet of paper with a 20 degrees angle, and the orientation detection is not (yet) smart enough to fix that, and it's just simpler to rescan the page.

Here is how I see things and why:

I think there should be a "properties" button and a button "delete" beside every document in the document list (like you did in your mockup). Now, having all these buttons would be ugly and kind of flood the user eyes in a bunch of useless visual elements.
So these buttons should only be displayed when the mouse goes over the document, or when the document is selected (--> works with touchscreens).

Same goes for the page grid: "edit" and "delete" should be on every pages. However to not flood the user and hides page details, they should only be displayed when the mouse goes over.

Problem is, when we select a page in the grid, it will switch to page list mode. So as-is, it doesn't work with touchscreens. I guess it can be solved by adding these two buttons also in the list mode. We can put them beside each page to not hide details of the page.


Also, I would like to point out that you focus too much on document titles IMHO. I'm not even sure I want to implement them.
Here is why: Paperwork goal is "scan & forget". In other words: Put all your documents in a big virtual stack, and let the computer do the job.
If we make it possible to change document titles and sort documents by titles, people will feel that, if they want to use Paperwork efficiently, they *have* to put a title on every documents. That's simply not true (or I did my job really poorly ...), and that would actually be using Paperwork *less* efficiently. Based on my experience, thumbnails are better than titles (except you can't sort them, but it hasn't been a problem yet ; hmmm... random idea: allow sorting thumbnails by image similarities ... ?)
Also, my main reason : it's one of those features that I won't use (and so won't test it much, and it will be buggy).

**Comment by MathieuJourdan** [comment]
Ok, thanks for explaining.

> I think there should be a "properties" button and a button "delete" beside every document in the document list (like you did in your mockup).

What about using the date (forget about the title for now) in the headerbar to edit properties, as shown in the mockup above ?

> So these buttons should only be displayed when the mouse goes over the document, or when the document is selected (--> works with touchscreens).

Seems good to me.

> Problem is, when we select a page in the grid, it will switch to page list mode. So as-is, it doesn't work with touchscreens. I guess it can be solved by adding these two buttons also in the list mode. We can put them beside each page to not hide details of the page.

Maybe we can use some sort of "selection mode" in grid view to allow dragging multiple pages to an other document or deleting them. "edit" would only be available in page list.

> Also, I would like to point out that you focus too much on document titles IMHO. I'm not even sure I want to implement them.

Yeah, that's a controversial point I wanted to introduce.

> people will feel that, if they want to use Paperwork efficiently, they have to put a title on every documents

Hum. Didn't think of that. This is absolutely true on the mockups.

> Based on my experience, thumbnails are much better than titles for identifying them

Actually I see this more as a "subject" than a title, when thumbnails fail to give a clue.

> Also, my main reason : it's one of those features that I won't use (and so won't test it much, and it will be buggy).

Really, that's probably nothing more than showing the 20 first caracters of the extra words in the papers list and the headerbar.

I'll iterate some more.

**Comment by jflesch** [comment]
> What about using the date (forget about the title for now) in the headerbar to edit properties, as shown in the mockup above ?

Possible. But there are 2 things that disturb me with this approach:
* there are no visual clue that the labels can be changed by clicking on the title/date
* shouldn't the option to set labels and change doc properties be closer to the doc and its labels ?

> Maybe we can use some sort of selection mode in grid view to allow dragging multiple pages to an other document or deleting them. "edit" would only be available in page list.

Hm, do you often need to handle multiples pages at once in Paperwork ?

> Actually I see this more as a "subject" than a title, when thumbnails fail to give a clue.

Yeah, that's one of the reason I made the thumbnails so big. But still, I think we can reduce them and keep most documents recognizables (labels actually help in that matter).

> Really, that's probably nothing more than showing the 20 first caracters of the extra words in the papers list and the headerbar.

Extra words are only here to palliate cases where the OCR doesn't work. For instance, I have some invoices with a crappy yellow background on which Tesseract didn't work. But even then, it's only useful for about 2 documents out of 989.

**Comment by mjourdan** [comment]
> Possible. But there are 2 things that disturb me with this approach:

> - there are no visual clue that the labels can be changed by clicking on the title/date
> - shouldn't the option to set labels and change doc properties be closer to the doc and its labels ?

I agree it is [not obvious](https://raw.githubusercontent.com/mjourdan/gnome-design/master/paperwork/paperwork-properties-empty.png) that the document date will allow the user to add labels, but I would argue that the dropdown icon make it [discoverable enough](https://raw.githubusercontent.com/mjourdan/gnome-design/master/paperwork/paperwork-properties.png).

If we place the edit button in the papers list, I'm afraid it turns one of the following:
- we have edit and delete icons horizontaly aligned, which results in very short width for thumbnails, text and labels
- we have edit and delete icons verticaly aligned, which would make popovers a bit obtrusive I think, but [why not](https://raw.githubusercontent.com/mjourdan/gnome-design/master/paperwork/paperwork-properties-left.png)?
- we have a dummy icon to open a popover to chose between edit and delete, which would lead to a very disruptive [dialog](https://developer.gnome.org/hig/stable/dialogs.html.en)

Perhaps we could imagine something sliding out under the item directly in the papers list, but that would require to code specific widget, wouldn't it? The bright side is it would allow to open a calendar as a popover to chose the date, which would be nice stuff I think.

> Hm, do you often need to handle multiples pages at once in Paperwork ?

Perhaps as much often as I open Paperwork for importing files. Use case: user has a lot of papers to scan. She brings a dead tree or two at work, and she scans from feeder to her email, as a single batch. At home, she imports that in Paperwork then uses split the file into several documents. This would be used from time to time, but could save a lot of time. That's the kind of [feature](https://github.com/jflesch/paperwork/issues/132) wich allows to spend less time both with paper and with software.

> Yeah, that's one of the reasons I made the thumbnails so big. But still, I think we can reduce them and keep most documents recognizables (labels actually help in that matter).

Sure. 

> Extra words are only here to palliate cases where the OCR doesn't work. For instance, I have some invoices with a crappy yellow background on which Tesseract didn't work. But even then, it's only useful for about 2 documents out of 989.

Sure, again. Displaying a few words when available seemed to me a nice answer to [make the extra text visible](https://github.com/jflesch/paperwork/issues/195), completing thumbnails without adding one-shot labels. But eh, if you're convinced that's a bad idea, I won't insist more and will make it disappear for the next mockups.

To do next:
- a screen for empty papers
- a simple label editor
- wires to attach labels to a paper

**Comment by jflesch** [comment]
>  I would argue that the dropdown icon make it discoverable enough.

I  think it's quite small for something used that often.

>  we have edit and delete icons horizontaly aligned, which results in very short width for thumbnails, text and labels
> we have edit and delete icons verticaly aligned, which would make popovers a bit obtrusive I think, but why not?

I was thinking about putting them vertically. I don't know if the popover would be obtrusive, but I must admit it feels weird to have it pop at variable locations.

> we have a dummy icon to open a popover to chose between edit and delete, which would lead to a very disruptive dialog

Ney.

> Perhaps we could imagine something sliding out under the item directly in the papers list, but that would require to code specific widget, wouldn't it? The bright side is it would allow to open a calendar as a popover to chose the date, which would be nice stuff I think.

This might work (and don't worry about the custom widgets).

Another possibility : Like for the export, we can use an info bar at the top of the page/grid view. (I'm really tired right now, so I'm not sure it makes sense)

> she scans from feeder to her email, as a single batch. At home, she imports that in Paperwork then uses split the file into several documents. This would be used from time to time, but could save a lot of time. That's the kind of feature wich allows to spend less time both with paper and with software.

Hm, it makes sense. It can be useful. However, I think we should keep in mind this is a rare use case, so I think this selection mode should be discreet.

> But eh, if you're convinced that's a bad idea, I won't insist more

:)
Just so you know, this ticket has officially became [my main example for a good feature request](https://github.com/jflesch/paperwork/wiki/FAQ#why-did-you-do-x-instead-of-y-).
(and yes, right now, I'm convinced)

**Comment by jflesch** [comment]
9a7d13cf9f775260061b98babb01b7747fa92512 : Display doc list + labels (still missing: thumbnails + buttons)

**Comment by jflesch** [comment]
a2fcc462672f052eee8c4c549bf674fee2c786a0 : Thumbnails in doc list

**Comment by jflesch** [comment]
7587645c002f476c7f234e4eabbe8c71b2261a19 : Implement custom progress bar (so it may disappear when unused)

**Comment by jflesch** [comment]
1a541aa1dbd0df4013b27597ca36d4d0a96d63ce : Doclist: add edit and delete buttons

**Comment by mjourdan** [comment]
> I think it's quite small for something used that often.

Actually dropdown icon is a very common trick, used by thousands of well-known websites and applications, including gmail, github, pocket, owncloud, and so on.

I made some more experiments:
- [compact properties popover](https://github.com/mjourdan/gnome-design/blob/master/paperwork/experiments/properties-list-popover-compact.png)
- [empty headerbar popover](https://github.com/mjourdan/gnome-design/blob/master/paperwork/experiments/properties-headerbar-empty.png)
- [selection widgets to edit directly from list](https://github.com/mjourdan/gnome-design/blob/master/paperwork/experiments/selection.png)

You'll notice I removed the additional keywords from the papers list, so editing them together with date is a bit surprising. I don't know what is the way to go.

**Comment by jflesch** [comment]
> Actually dropdown icon is a very common trick, used by thousands of well-known websites and applications, including gmail, github, pocket, owncloud, and so on.

Yes, but from an ergonomic standpoint, there are quite small targets to click on with a mouse. Which means they should be used only for operations that are not done often. But changing the labels is actually quite common in Paperwork.

> selection widgets to edit directly from list

I think this idea is interesting. It avoids pop-over, which, IMHO, is nice.

I think that when the user clicks on "edit", we can extend the space allocated to the document in the list. This way, we can make label selection faster (maybe like the label list existing currently in Paperwork 0.2 ?).

We can also add an a button below the labels that would open a dialog to type in the additional keywords (again, extra keywords are and should rarely be used, so it's not a big deal if it's hard to reach).
What do you think ?

**Comment by jflesch** [comment]
6ab296400f43f084856a22762977b67a9894b241 : Popover dialog for view settings (not connected yet)

**Comment by mjourdan** [comment]
> Yes, but from an ergonomic standpoint, there are quite small targets to click on with a mouse.

Nah, the icon is only here to advertise there is something to activate, the actual target is the icon + the item besides it. In our case, that would be the whole headerbar title.

> I think this idea is interesting. It avoids pop-over, which, IMHO, is nice.

> I think that when the user clicks on "edit", we can extend the space allocated to the document in the list. This way, we can make label selection faster (maybe like the label list existing currently in Paperwork 0.2 ?).
> We can also add an a button below the labels that would open a dialog to type in the additional keywords (again, extra keywords are and should rarely be used, so it's not a big deal if it's hard to reach).
> What do you think ?

Both points seem right. We even could use a simple ">" as the only button for each item. Activating it would slide the whole left pane, with the following (sorry, can't do mockups for now):
- "<" in the headerbar to return to the papers list
- a bin icon in the headerbar (in place of the "+") to delete the files
- in the properties pane, a calendar widget (eventually in a popover) + text field to change the date
- the list of labels (scrollable if too long) + text field
- at the bottom, the additional keywords text area (eventually expandable)

**Comment by mjourdan** [comment]
Please find [some wires](https://github.com/mjourdan/gnome-design/blob/master/paperwork/wires/first-paper.png) which cover paper creation and properties edition.

**Comment by jflesch** [comment]
1) welcome popup: KO (I *hate* useless popups)
2) "No paper yet !": OK
3) "Use the + button to add your first paper": OK
4) "New document" that appear only when we click on "+" : OK
5) "Use the scan button (...)": OK
6) Scan button locked: OK ; Maybe we can turn it into "Cancel" to stop the scan ? (may be useful if the paper is not correctly inserted in the scanner)
7) Progress bar on top : Actually, I put it at the bottom for now. At the top, it raises the question : should everything move down, or should the progress bar hides something ?
8) The panel to edit document properties : Looks good. However, if we want it to be clear, I need to be able to animate the switch from the document list to the document properties (without animation, I think the switch may be disturbing for the user). I'm not sure if there is an easy way to do that with Gtk.

**Comment by mjourdan** [comment]
1) Useless indeed, would fit an app center, though
7) I can't see why the question would not be raised at the bottom
8) Absolutely; I was thinking of slide effect similar to what popovers provide. [GtkRevealer](https://developer.gnome.org/gtk3/stable/GtkRevealer.html), maybe ?

**Comment by jflesch** [comment]
7) because at the bottom, moving the content up would seem weird --> it can only hide stuff. Also, usually, less important information are at the bottom --> less risk of hiding something important.
8) Ok. I'll have to try it.

**Comment by jflesch** [comment]
8) Tried and adopted

Just one thing : on your latest mockups, when editing a document, there is no way to cancel the changes. So currently I made things slightly differently : In the panel used for editing, there are at the bottom buttons 'ok' and 'cancel'.

**Comment by jflesch** [comment]
Hm, screw the cancel button. I'll do it your way

**Comment by jflesch** [comment]
GtkRevealer : 99adc2a77d1803e658cef52a1ac54ba26814d0e8 b0886a9fc58a117ec8af7b31c160ea8efc3655de ff543206aca56b1ebc1968a75148793fa44cfb41

**Comment by tYYGH** [comment]
jflesch, mjourdan, I’m impressed with the mockups, the quick adoption, and real improvements this brings!

I post this only to comment about tags. While I agree with most of mjourdan’s improvements, he/she seems to relegate tags to second-class objects in Paperwork, which may explain the focus on titles…
Of course, each person may use the software in a slightly different way, but IMHO tags are important, much like they can be on a blog or an email client software. (BTW, +1 for #362!)

I like a lot the idea to start typing a tag and have it auto-completed and changed into an on-screen label with a delete-button. I have a question: what happens when several tags start the same? For instance, since one may not remember the exact spelling of a tag, it would be nice to have a list open just like in the web browser’s search bar, with all tags matching “*typed text*”, case-insensitive and disregarding diacritics.

**Comment by jflesch** [comment]
Actually, right now, I took an approach similar to before : http://jflesch.kwain.net/~jflesch/paperwork/paperwork_redesign_progress2.png

Why ? Because I'm lazy, and it was the fastest to implement for now :p

Anyway, I think Mjourdan approach is interesting :
- advantage : if there are a lot of labels, the UI is not drowned
- disadvantage : users have to remember which tags exist and which don't

When using Paperwork, the goal is to be lazy. So I'm going to assume that users don't have that many labels, so drowning in the mess shouldn't be an issue.
However, having to remember what labels are available is not exactly what I call "the lazy way".

What do you think ?

**Comment by tYYGH** [comment]
Some context first: I have 697 documents, and I currently use 51 tags, ranging from 5 tags being used only once so far, to 5 tags being used more than 150 times (max 338). On average 3 tags per document, which is consistent with my usage: tags for “who”, “what”, “where”.

In my opinion, remembering the labels is not that much of an issue: I may not remember the casing, or the presence or not of accents, or the presence or not of an article before the noun, or the order of words in multi-word tags… but I doubt the number of tags will become such that remembering the key words will be a problem.
My wish to have a suggestion-list along with auto-completion would answer two needs:
— the need to remember spelling, eg: I type “kid” and it provides me with “livret €kid”,
— the need find out existence, eg: I type “Fnac” and it provides me with nothing, meaning no tag exists with this word.
There would need to be some place where all tags can be seen, though, for actions such as deleting obsolete tags, or renaming them.

That being said, the current presentation as a full list is OK as far as I’m concerned, even though my list is longer than the window ;-)

**Comment by jflesch** [comment]
> That being said, the current presentation as a full list is OK as far as I’m concerned, even though my list is longer than the window ;-)

Don't worry. A scrollbar will come to your rescue :)

**Comment by mjourdan** [comment]
On 13/02/2015 10:30, tYYGH wrote:
> jflesch, mjourdan, I’m impressed with the mockups, the quick adoption, and real improvements this brings!

Thanks!

> I post this only to comment about tags. While I agree with most of mjourdan’s improvements, he/she seems to relegate tags to second-class objects in Paperwork, which may explain the focus on titles…
> Of course, each person may use the software in a slightly different way, but IMHO tags are important, much like they can be on a blog or an email client software. (BTW, +1 for #362!)
>
> I like a lot the idea to start typing a tag and have it auto-completed and changed into an on-screen label with a delete-button. 

I think there is a little misunderstanding here. When you talk about
second-class objects, I guess you refer to
[this](https://github.com/mjourdan/gnome-design/blob/master/paperwork/wires/first-paper.png)
where user types in two labels. As the wires cover the very first paper
digitalization, there are absolutely no existing label to be shown. Is
this your concern?

So, when jflesch wrote "- disadvantage : users have to remember which
tags exist and which don't", this is true only on first start.

However this gives me an idea. To invite people to use labels, and make
the properties dialog less surprising, we could apply a dummy label for
the very first document. Would it help?

I would also like to point [this
mockup](https://github.com/mjourdan/gnome-design/blob/master/paperwork/paperwork-label-filter.png)
out.

I have a question: what happens when several tags start the same? For
instance, since one may not remember the exact spelling of a tag, it
would be nice to have a list open just like in the web browser’s search
bar, with all tags matching “*typed text*”, case-insensitive and
disregarding diacritics.

Well, this is indeed the intended behavior, I think I sketched that on
paper but I didn't go further with inskcape.

Regards

**Comment by mjourdan** [comment]
Here are some [wires](https://github.com/mjourdan/gnome-design/blob/master/paperwork/wires/selection-mode.png) covering move, edit and delete onto multiple pages (see #132).

**Comment by jflesch** [comment]
Hm, looks good, but I have some questions :
- Do you think multiple selection is really useful ?
- In the edition buttons, you suggest a button to mirror the page. Is there a use case for it ?
- I don't get how the user is supposed to switch from the page list mode to the page grid + multiple selection mode ?

FYI, for now, when opening a document, if it has only one page, the page list mode is used by default. If it has many pages, the page grid mode is used. A single click on a page allows to switch from grid mode to list mode.

Also, just a quick note regarding cropping : It can only be done on one page at a time. Also, it requires precision, so I don't think it makes sense to do it from page grid mode.

**Comment by jflesch** [comment]
> So, when jflesch wrote "- disadvantage : users have to remember which tags exist and which don't", this is true only on first start.

Ok, I think I misread your wires. But does typing a label name allow to create a new label, or does it allow to filter the label list ?

**Comment by mjourdan** [comment]
> Do you think multiple selection is really useful ?

Yep. See the use case described [here](https://github.com/jflesch/paperwork/issues/356#issuecomment-67692268) and [there](https://github.com/jflesch/paperwork/issues/356#issuecomment-70557040), plus the one below.

> In the edition buttons, you suggest a button to mirror the page. Is there a use case for it ?

Yeah, actually page orientation detection never worked well for me, and usually give me bad results (while OCR does work well). As it is very long to proceed, I disabled it to save time. But my scanner has a lid and I never remember where I should put the top of the page. In both cases, I end up with pages upside-down, far most frequently than I get portrait position instead of landscape.

> I don't get how the user is supposed to switch from the page list mode to the page grid + multiple selection mode ?

Yeah, sorry, I did'nt show that. It's by activating the check mark button. You may want to take a look at the [gnome hig](https://developer.gnome.org/hig/stable/selection-mode.html.en) to decide if it is appropriate for Paperwork or not.

**Comment by mjourdan** [comment]
> Ok, I think I misread your wires. But does typing a label name allow to create a new label, or does it allow to filter the label list ?

Both. It allows to keep only matching labels from the list, and create a new one (with a pre-filled color) when there is no match. I hope to make more detailed wires tomorrow.

**Comment by jflesch** [comment]
> Yeah, actually page orientation detection never worked well for me, and usually give me bad results (while OCR does work well). As it is very long to proceed, I disabled it to save time. But my scanner has a lid and I never remember where I should put the top of the page. In both cases, I end up with pages upside-down, far most frequently than I get portrait position instead of landscape.

With Paperwork <= 0.2.2, make sure you have the correct dictionary installed (if you're French, aspell-fr for instance, but it should also work with myspell-fr and ispell-fr I think). It's required to get correct results.
With Paperwork which is currently in the unstable branch, Tesseract orientation detection will be used. It gives much much better results :)

Anyway, if I understand correctly, this icon is supposed to mean "rotation of 180 degrees", right ? Because as-is, I understand it as "vertical mirroring", which is not the same thing.


> Yeah, sorry, I did'nt show that. It's by activating the check mark button. You may want to take a look at the gnome hig to decide if it is appropriate for Paperwork or not.

I didn't realized it is a mode disabled by default. OK for me (... and more work for me ...)

**Comment by mjourdan** [comment]
> Also, just a quick note regarding cropping : It can only be done on one page at a time. Also, it requires precision, so I don't think it makes sense to do it from page grid mode.

Use case: I forgot to calibrate my scanner (actually I didn't understand what "calibrate meant" when I saw it in the preferences window), so every page should be equally cropped. On the other hand, your are right about precision.

> With Paperwork <= 0.2.2, make sure you have the correct dictionary installed (if you're French, aspell-fr for instance, but it should also work with myspell-fr and ispell-fr I think). It's required to get correct results. With Paperwork which is currently in the unstable branch, Tesseract orientation detection will be used. It gives much much better results :)

I got aspell-fr and myspell-fr-gut installed among others. I'll probably wait 0.3 is ready before I enable this again, then.

> Anyway, if I understand correctly, this icon is supposed to mean "rotation of 180 degrees", right ? Because as-is, I understand it as "vertical mirroring", which is not the same thing.

Ah. It is meant as a 180° rotation, as in gnome display settings. Indeed this icon is also used by inkscape for instance as vertical mirroring. Don't know if we could find a less confusing icon.

> I didn't realized it is a mode disabled by default. OK for me (... and more work for me ...)

And less work for users \o/

**Comment by jflesch** [comment]
> Ah. It is meant as a 180° rotation, as in gnome display settings. Indeed this icon is also used by inkscape for instance as vertical mirroring. Don't know if we could find a less confusing icon.

I think with Tesseract's orientation detection, there will be much less need for page rotation. Also, clicking instead 2 times on "rotate right" (or left) isn't a problem. So I think we can safely get ride of this button.

**Comment by mjourdan** [comment]
This evening I started working on [wires for browsing by labels](https://raw.githubusercontent.com/mjourdan/gnome-design/master/paperwork/wires/browsing-labels.png), so I didn't have time to make wires for labels creation / edition. Sorry.

**Comment by jflesch** [comment]
Wouldn't make it more sense to attach the popover to the search field instead of the header bar ?

On an unrelated note, I'll be working on another personnal project for the next few months. I'll keep replying here, but I don't think I'll have much time to code on Paperwork :/

**Comment by mjourdan** [comment]
> Wouldn't make it more sense to attach the popover to the search field instead of the header bar ?

I don't know if it would make *more* sense. What I mocked up was "browse every papers that belong to this label (and eventually later search for papers inside this restricted set)". If we attach the popover to the search field, we would have "search only for papers that belong to all selected labels (but do search or the list won't be filtered out)".

> On an unrelated note, I'll be working on another personnal project for the next few months. I'll keep replying here, but I don't think I'll have much time to code on Paperwork :/

Ok, thanks.

**Comment by mjourdan** [comment]
Hey,

Here are some more wires to attach labels to a paper:
- [pick up existing labels](https://raw.githubusercontent.com/mjourdan/gnome-design/master/paperwork/wires/labels-list.png)
- [typing](https://raw.githubusercontent.com/mjourdan/gnome-design/master/paperwork/wires/labels-typing.png)

**Comment by mjourdan** [comment]
Here is a review of what I found in the repository. I hope to send some patches, at least for the easy items.

### WIP Review

Paper view:

- headerbar title is odd (empty GtkLabel makes title consume two lines)
- page selector is too wide (up to 4 digits while 2 should be enough)
- zoom using a slider does not work
- quick scanner setup for first page scan (needs design)

Properties:

- create, edit and delete labels does not work
- use a calendar to change the date (needs design?)

Preferences:

- dialog is scary: offering 5 times for help makes me feel like I'm in distress
- dialog lacks headerbar (needs design)

Help:

- help is not available without network connection


### Annoying bugs

- install fails with "error: package directory 'src/paperwork/frontend/doceditdialog' does not exist"
- in paper list, controls (edit and delete) sometimes disappear
- with the first paper opened in grid view, clicking in an empty zone switches the view to a zoom on the second paper

**Comment by jflesch** [comment]
> install fails with "error: package directory 'src/paperwork/frontend/doceditdialog' does not exist"

This is a work in progress, do not install it ! :)
You can run it from the repository by running 'src/launcher.py'.

> with the first paper opened in grid view, clicking in an empty zone switches the view to a zoom on the second paper

Funny, I didn't see this bug. It's actually even weirder for me, because when I test it, it switches to another document.

**Comment by mjourdan** [comment]
> This is a work in progress, do not install it ! :)
> You can run it from the repository by running 'src/launcher.py'.

Stupid me!

> Funny, I didn't see this bug. It's actually even weirder for me, because when I test it, it switches to another document.

Yeah, by paper I meant document, not page.

**Comment by jflesch** [comment]
Ok, it's working good-enough for the unstable branch so I merged it and destroyed the branch 'wip-ui-redesign'.

Here is a review of what I found in the repository. I hope to send some patches, at least for the easy items.

### WIP Review

Paper view:

- [x] headerbar title is odd (empty GtkLabel makes title consume two lines)
- [x] page selector is too wide (up to 4 digits while 2 should be enough)
- [X ] zoom using a slider does not work
- [ ] quick scanner setup for first page scan (needs design)
- [x] doclist draws too much attention
- [ ] sometimes no paper is shown because none is selected in doclist
- [X] Page edition
- [X] Page deletion

Properties:

- [x] use a calendar to change the date
- [x] when changing the date of a paper, headerbar is not updated
- [x] 'add label' doesn't work
- [x] 'edit label' doesn't work
- [x] text is not vertically aligned in labels
- [x] the label list has no border
- [ ] the additional keywords text area has no border
- [x] what are the « additional keywords » for is not obvious
- [X ] labels listbox is thick (checkboxes should disappear, edit button should be replaced with the label color)

Label Editor:
- [x] transparency can be set but labels are always opaque
- [x] when editing a label, sometimes the text is right but the color is from the previous label
- [x] warnings like "drawing failure for widget `GtkBox': NULL pointer" (seem related to the color widget)
- [x] display is buggy when opening the color editor

Preferences:

- [x] dialog is scary: offering 5 times for help makes me feel like I'm in distress
- [x] dialog lacks headerbar (needs design)
- [x] dialog offers 3 way to close (Cancel, Ok, and close)

Help:

- [ ] help is not available without network connection (Jflesch> Won't do for 0.3.x)

### Annoying bugs

- [X ] install fails with "error: package directory 'src/paperwork/frontend/doceditdialog' does not exist"
- [ ] in paper list, controls (edit and delete) sometimes disappear
- [ ] with the first paper opened in grid view, clicking in an empty zone switches the view to a zoom on the second paper

**Comment by jflesch** [comment]
It has been a very long time since there was a major Paperwork release, and this GUI is starting to look really great :) --> I would like to ship a 0.3.0 asap. We can still fix minor issues later in the next minors releases (0.3.x).

Anyway, I have a few questions for you:

> quick scanner setup for first page scan (needs design)

What do you mean exactly ?

>  sometimes no paper is shown because none is selected in doclist

Well .. it makes sense I guess, no ? :)
I guess we could display the last used documents instead of nothing. But is it blocking for 0.3.0 ?

> the additional keywords text area has no border

I didn't notice it as being a problem.
Maybe it could be a problem based on the Gtk theme used. But then again, is it a Paperwork problem ?

> help is not available without network connection (Jflesch> Won't do for 0.3.x)

This one will require some work, and I'm fairly sure that current users of Paperwork all have an Internet connection :). However, I must admit it's a problem because it makes Paperwork dependent on GitHub, and pages can't be translated. Still, I think it would be best to keep this change for 0.4.0.

> in paper list, controls (edit and delete) sometimes disappear

Do you have any reproduction scenario ? I can't find any anymore :(

> with the first paper opened in grid view, clicking in an empty zone switches the view to a zoom on the second paper

Can't reproduce. However, I did a lot of changes, so I may have fixed it by accident :-). Can you have a look please ?


So, what do you think ?

**Comment by tiramiseb** [comment]
Hello,

I'm currently trying this new UI. Some comments:

* multiple pages preview is a great idea but not very effective as-is: pages previews are generated given a size (and I can change this size, great), but I would prefer setting a "number of pages on a single row", pages size would change when resizing the window
* margin between pages is way too large
* I would prefer having three buttons instead of a dropdown menu on the "scan" button. maybe a dropdown on the side as in Simple Scan...

**Comment by jflesch** [comment]
> multiple pages preview is a great idea but not very effective as-is: pages previews are generated given a size (and I can change this size, great), but I would prefer setting a "number of pages on a single row", pages size would change when resizing the window

I understand your idea, but I'm not sure how it can be presented without making the GUI really complex.

> margin between pages is way too large

In which view ? Grid, list, or both ?

>I would prefer having three buttons instead of a dropdown menu on the "scan" button. maybe a dropdown on the side as in Simple Scan...

Hm, good point

**Comment by tiramiseb** [comment]
> I'm not sure how it can be presented without making the GUI complex to use

By default, multi-page documents would display 3 pages per row, no matter the size of the page. The same slider than now would be used to set the number of documents per row. But instead of setting the size of a single document, it would set the number of documents per row (with gaps between two steps on the slider ; eg. settings for number of cpus or number of screens in virtualbox).

Problem with fixed thumbnails sizes is that if the window is 1 pixel narrower than a number of thumbnails, the last thumbnail does on the next line and there is a wide empty space.

> see for instance Evince, when it's freshly started and display the last documents used ; or any file browser

Displaying pages in a document and files in a directory is not the same thing. content vs icons.

> In which view ? Grid, list, or both ?

grid. the space between thumbs in the grid (which can be pretty small) is the same as the space between pages in the list (which is usually large, 100% width zoom by default ?).

BTW, in all situations Ctrl + mouse wheel should changes pages size, on-the-fly.

**Comment by jflesch** [comment]
> By default, multi-page documents would display 3 pages per row, no matter the size of the page.

Why 3 ? :)
Also, a problem with that is that pages can have various width. You can have portrait pages on the first line, and only landscape pages on the second line. The zoom level wouldn't be the same for both lines.

So I just implemented something a little bit different :
Paperwork checks if it can fits all the pages in one line without making the previews too small. If it can, it does, if not, it fall back to the previous grid layout. It doesn't really solve the problem for big documents, but it does for small ones.

Please tell me what you think of it ?

> grid. the space between thumbs in the grid

I reduced the margin. Again, please tell me if it is ok for you.

> in all situations Ctrl + mouse wheel should changes pages size, on-the-fly (Ctrl + "+" and Ctrl + "-" too).

Good point. I'll see what I can do

**Comment by jflesch** [comment]
Hm actually, maybe you're right.
In grid mode, the slider would be the number of pages par line
In page mode, the slider would be the level of zoom (as of now).

I can give it a try this evening

**Comment by tiramiseb** [comment]
> Paperwork checks if it can fits all the pages in one line without making the previews too small. If it can, it does, if not, it fall back to the previous grid layout

I really didn't notice that at first. It's not clear.
And there is a problem: with document with few pages (eg. 4 pages), there may be a large empty space below them. I have a document with 4 pages, the space below them is higher than the height of a single page.
But it would be the same if I manually selected the pages number of sizes.

> I reduced the margin

It's better

> Hm actually, maybe you're right.

Yes I am ! I am always right !!! :smiley: 

However, when there are less pages than the number of pages configures, they could be larger than the selected value...


----

After playing more with that, I realized something : the current slider is in a drop-down menu, displaying with an annoying animation. It makes me feel like it is a long-term setting which should not be used ofter.
But in fact it is only a zoom level slider. A zoom level slider should either be always or never visible.
Example for "always" : LibreOffice, low-right corner
Example for "never": Evince, textual percent drop-down / edit box

The same applies to switching between "page" and "grid" views. Buttons on a popup doesn't feel right. 2 buttons on the bar would be better.


Moreover, the "highlight words" is more like a long-term setting and should not be placed with zoom parameters.

**Comment by jflesch** [comment]
While I agree that the slider could control the number of pages per line, I disagree with moving the settings.

The current logic is the same for the zoom level, the grid/list layout buttons and the "highlight words" switch :
- They all control the way pages are displayed, and so, IMHO should remain together
- In a perfect world, those are things you should never have to use

The zoom level is always annoying to set correctly (LibreOffice and Evince actually gives good examples: if often end up being trial-and-errors)
The layout : If there is only one page, the layout 'grid' is useless anyway. If there are more, you start with the grid, and then look at one or two pages specifically. I'm not sure there are so many cases where you have to switch back to the grid layout.
"Highlight words" : That's more a debug feature than a really useful one. I guess it helps people relax by making sure the OCR worked (I know it does for me ;) ).

So in the end, if the UI is done right, you shouldn't have to use them really often.

Also, I think moving them out of this dialog would clutter the main window.

**Comment by tiramiseb** [comment]
Understood.
In this case, I think this behaviour is a step towards a UI done right / a perfect world.
It should be used by as many people as possible to check which improvements could be done.
I will use it and see if it fits my usage of Paperwork.

**Comment by jflesch** [comment]
Note to myself: I must also re-add support for drag'n'drop also (the API tends to be really painful to use ; I did some crappy work in the previous UI :/)

**Comment by tiramiseb** [comment]
I don't know if I should open a new issue for the following thought...

Clicking on "New document" on the documents list displays an empty document (zero page)
Clicking on "+" makes the documents list jump to "New document", displaying zero page
Clicking on "Scan" gives a list of options for importing documents (scan or pdf import). Even if a document is displayed, the scan (or import) is done in a new document.

My idea:
- "New document" and "+" are useless
- "Scan" can be split in 3 buttons (or a button for "default action" + a dropdown), placed on the left panel in place of the "+" button

I think "Scan" has nothing to do with the current document, nothing to do on the right panel.

**Comment by jflesch** [comment]
> "New document" and "+" are useless

How would you start making a new document then ? oO

> I think "Scan" has nothing to do with the current document, nothing to do on the right panel.

"Scan" add a scanned page to the *current* document. Always.

**Comment by tYYGH** [comment]
Maybe scan and import just shouldn't stay together: import would go to the left, and scan would remain to the right.

**Comment by tYYGH** [comment]
Unless of course you intend to allow importing additional PDF content to an existing document…

**Comment by jflesch** [comment]
I don't.

But as of now, "import" can import PDF as a document, or images as extra pages to the current document :/

**Comment by tiramiseb** [comment]
Wrong memory on my side, sorry, you are of course right. Recently I have only imported PDF documents :)

However the behaviour of the "Scan" button is not consistent, as you have said.

Another suggestion:
* the "+" button has the "scan" drop-down menu, to allow directly scanning to a new document or importing a new file
* the "scan" button is only displayed on existing images-based documents which allow new pages to be scanned (hidden or grayed out on PDF documents)
* the "new document" entry doesn't need to exist

**Comment by mjourdan** [comment]
> I don't know if I should open a new issue for the following thought...

Seems to me you're far from being off topic.

Let me rephrase you're idea to get this straight. What you are saying, tiramiseb, is it is nonsense to create document of 0 page, therefore documents should always be created either from file or scanner. Then it is allowed to add pages when a document has been created from scanner, when you imported an image file. However, if you imported a pdf, it is not allowed, because a pdf is considered to be a whole document. Am I right?

Let say a user does create a document using some .jpg file. Would her be able to add pages from her scanning device, and why?

**Comment by tiramiseb** [comment]
Regarding behaviour when it is a scanned document or a PDF file or an image import, I'm keeping the existing behaviour.

Globally, my idea is to totally remove the "document of 0 page" and to be able to:
- "scan or import a new document" on the left panel
- "scan or import into the current document" on the right panel

However, the current behaviour being "a PDF-based document cannot be modified", the "scan [...]" on the right panel would only be available for scanned docs or images.

**Comment by mjourdan** [comment]
Le 07/10/2015 11:57, Jerome Flesch a écrit :
> It has been a very long time since there was a major Paperwork release, and this GUI is starting to look really great :) --> I would like to ship a 0.3.0 asap. We can still fix minor issues later in the next minors releases (0.3.x).
> 
> Anyway, I have a few questions for you:
> 
>> quick scanner setup for first page scan (needs design)
> 
> What do you mean exactly ?

I launch Paperwork, start my scanner, and try to scan a document.
Nothing happens, and I have to open the Preferences so that my device is
detected. This happens every time I launch Paperwork, and is
uncomfortable, as my device is the same for years now. At least, that
was true by the time I wrote this.

> 
>>  sometimes no paper is shown because none is selected in doclist
> 
> Well .. it makes sense I guess, no ? :)
> I guess we could display the last used documents instead of nothing. But is it blocking for 0.3.0 ?

Yeah, my point was we should always have a document selected, so that we
have something to show, whether it is the last used, or a new empty doc
(in which case we would have a nice message saying "this doc is empty,
you can add pages blah blah"). While it is confusing, I wouldn't say
it's a blocking issue.

>> the additional keywords text area has no border
> 
> I didn't notice it as being a problem.
> Maybe it could be a problem based on the Gtk theme used. But then again, is it a Paperwork problem ?

I took a look at a couple of apps (polari, boxes...) and I couldn't
figure out how it happens I can't set a stupid border on this particular
widget. But really it's just about polish.

>> help is not available without network connection (Jflesch> Won't do for 0.3.x)
> 
> This one will require some work, and I'm fairly sure that current users of Paperwork all have an Internet connection :). However, I must admit it's a problem because it makes Paperwork dependent on GitHub, and pages can't be translated. Still, I think it would be best to keep this change for 0.4.0.

Seems wise to me.

>> in paper list, controls (edit and delete) sometimes disappear
> 
> Do you have any reproduction scenario ? I can't find any anymore :(

As far as I remember, I couldn't identify a pattern to reproduce, but
this definetely happened multiple times!

>> with the first paper opened in grid view, clicking in an empty zone switches the view to a zoom on the second paper
> 
> Can't reproduce. However, I did a lot of changes, so I may have fixed it by accident :-). Can you have a look please ?
> 
> 
> So, what do you think ?

I'll test a bit and give some feedback soon.

**Comment by mjourdan** [comment]
> Globally, my idea is to totally remove the "document of 0 page" 

This could be a good thing indeed!

> - "scan or import a new document" on the left panel
> - "scan or import into the current document" on the right panel

I'm not sure about this. This means user would have to chose the source in two different places instead of one, which would result imho in a more complex experience.

**Comment by jflesch** [comment]
> I launch Paperwork, start my scanner, and try to scan a document. Nothing happens, and I have to open the Preferences so that my device is detected. This happens every time I launch Paperwork, and is uncomfortable, as my device is the same for years now. At least, that was true by the time I wrote this.

I think this one is a bug unrelated to the UI changes. The expected behavior, even in stable:
- if Paperwork was never configured, open the configuration dialog
- if it has already been configured (even in a previous session), just do the scan

> Yeah, my point was we should always have a document selected, so that we have something to show, whether it is the last used, or a new empty doc (in which case we would have a nice message saying "this doc is empty, you can add pages blah blah"). While it is confusing, I wouldn't say it's a blocking issue.

[Noted, but for 0.3.x](https://github.com/jflesch/paperwork/issues/399)

> As far as I remember, I couldn't identify a pattern to reproduce, but this definetely happened multiple times!

Actually, in the meantime, I reproduced it and fixed it :) (please tell me if you see it again)

> I'm not sure about this. This means user would have to chose the source in two different places instead of one, which would result imho in a more complex experience.

I agree. The current way is messy and unperfect, but obviously most users (if not all) found their way. I doubt we will find a perfect solution anyway. So I think we will just let it as it is for 0.3.0.

I'll just split the "scan" button in two --> "scan" button + a button to have the sub-menu, as done before in 0.2.x

**Comment by tYYGH** [comment]
About the scanner and the preferences:
> I launch Paperwork, start my scanner, and try to scan a document.
> Nothing happens, and I have to open the Preferences so that my device is detected.
> This happens every time I launch Paperwork, and is uncomfortable, as my device is
> the same for years now. At least, that was true by the time I wrote this.

Same for me. Probably because it is an USB scanner, that I plug only when I need it, and it would get a different system ID each time…

**Comment by jflesch** [comment]
My scanner is also a usb one (HP), but I never unplug it. Anyway, it's still a bug in that case : if the scanner id cannot be found, it should either display a popup saying "scanner not found" or open the settings window.

**Comment by tYYGH** [comment]
About the “empty document” thing:
> Yeah, my point was we should always have a document selected, so that we have
> something to show, whether it is the last used, or a new empty doc (in which case
> we would have a nice message saying "this doc is empty, you can add pages blah
> blah"). While it is confusing, I wouldn't say it's a blocking issue.

I don’t agree. I find nothing either unnatural or confusing about seeing an empty document. In the real world, there are empty folders; there are empty pages; there are empty notebooks… However, it is a really good idea to have a message like “This is a new, empty, document. You can import or scan new pages inside.”

As for always having a document selected, I even more don’t agree. When I start Paperwork, it is:
— either because I will scan/imports new documents, in which case being on a new, empty, document is just perfect;
— or because I have a search to do, and then I don’t care at all what document was selected last time.

Cheers,

**Comment by tYYGH** [comment]
rel. scanner: I do have a “scanner not found” message. Maybe some heuristics using the scanner name instead of its ID if the ID is not found would solve the issue?

**Comment by jflesch** [comment]
Hm, I prefer not using heuristic when possible. They tend to be hard to debug (for instance, I'm fairly certain the label guessing has some bugs).

Here, I guess it can be done by simply looking how many scanners are connected. If there is only one ... well it must be this one.
I will just have to ignore camera devices. Since their IDs are prefixed with "v4l:", it should be easy.

Added: #403

**Comment by jflesch** [comment]
Implemented:
- Drag-n-drop: Page --> another document
- Drag-n-drop: Page --> another position in the same document

**Comment by jflesch** [comment]
Most of the GUI changes for 0.3.0 have been implemented (in the branch 'unstable'). I'm going to close this meta ticket.
Please feel free to open a new ticket for each issue you (still) see.
